OAuth2 token introspection panics and crashes the whole Authorino process on a non-200 / missing-`active` introspection response

**Describe the bug**

The OAuth2 token-introspection identity evaluator (`pkg/evaluators/identity/oauth2.go`) trusts the introspection endpoint's response unconditionally. It never checks the HTTP status code and uses a non-comma-ok type assertion on the `active` claim:

```go
resp, err := httputil.NewClient(oauth.Timeout).Do(req)   // no status-code check
...
if claims["active"].(bool) {   // oauth2.go:92 — panics if "active" is absent or not a bool
```

Per RFC 7662, `active` is only guaranteed on a **successful (200)** introspection response. When the introspection request itself is rejected; most commonly because the configured `clientSecret` is wrong or has been rotated, providers such as Keycloak return HTTP **401** with a *valid JSON* error body like `{"error":"invalid_request","error_description":"Authentication failed."}`. That body decodes fine (so the decode-error branch is not taken), `claims["active"]` is `nil`, and the assertion panics with `interface conversion: interface {} is nil, not bool`.

Because evaluators run inside goroutines spawned by the auth pipeline (`pkg/service/auth_pipeline.go`, `evaluateAuthConfigs`) and there is **no `recover()` anywhere in the codebase**, this panic is not contained to the one request.., it terminates the entire Authorino process (exit code 2). Since one instance serves ext_authz for every indexed host, a single misconfigured AuthConfig takes down authorization for **all** protected services, and then crash-loops as the offending request repeats.

**Help us Reproduce it**

1. Deploy Authorino + Envoy (as ext_authz) + Keycloak + a sample upstream (I used `make local-setup DEPLOY_KEYCLOAK=1` with the `kuadrant` realm from `kuadrant/authorino-examples`).
2. Create a Secret whose `clientSecret` is **wrong** (or rotate the real one so introspection is rejected):
   ```yaml
   apiVersion: v1
   kind: Secret
   metadata: { name: oauth2-introspection-credentials, namespace: default }
   stringData:
     clientID: talker-api
     clientSecret: WRONG-SECRET-triggers-401
   ```
3. Apply an AuthConfig using OAuth2 token introspection:
   ```yaml
   apiVersion: authorino.kuadrant.io/v1beta3
   kind: AuthConfig
   metadata: { name: talker-api-oauth2-introspection, namespace: default }
   spec:
     hosts: ["talker-api"]
     authentication:
       keycloak-opaque:
         oauth2Introspection:
           endpoint: http://keycloak.default.svc.cluster.local:8080/realms/kuadrant/protocol/openid-connect/token/introspect
           tokenTypeHint: access_token
           credentialsRef: { name: oauth2-introspection-credentials }
   ```
   (Status becomes `ready: true`.)
4. Send a normal request through Envoy with any bearer token:
   `curl http://localhost:8000/get -H "Host: talker-api" -H "Authorization: Bearer $ACCESS_TOKEN"`
5. Authorino calls the introspection endpoint, Keycloak responds `HTTP 401` with `{"error":"invalid_request","error_description":"Authentication failed."}`, and Authorino panics at `pkg/evaluators/identity/oauth2.go:92` and **the pod crashes** (`lastState.terminated.reason=Error`, `exitCode=2`, `restartCount` increments on every such request).

Observed Authorino log at the crash:
```
DEBUG authorino.service.auth.authpipeline.identity.oauth2  sending token introspection request ...
panic: interface conversion: interface {} is nil, not bool
github.com/kuadrant/authorino/pkg/evaluators/identity.(*OAuth2).Call
    /usr/src/authorino/pkg/evaluators/identity/oauth2.go:92
```
Envoy returns `403` to the client because the ext_authz connection drops mid-request (`failure_mode_allow` false). Directly confirming the trigger, the same introspection call returns `HTTP 200` with `"active"` present when the correct secret is used, and `HTTP 401` `{"error":"invalid_request",...}` (no `active`) with a wrong secret.

**Expected behavior**

Instead of panicking and crashing the process, a non-2xx or otherwise malformed introspection response (missing/non-boolean `active`) should result in a clean, per-request **denial** (`UNAUTHENTICATED`) with a descriptive reason. A misconfigured or transiently-failing introspection endpoint must never bring down the Authorino instance or affect other AuthConfigs.

**Environment (please complete the following information):**
- Cluster information: Kind v0.31.0, Kubernetes v1.35.0 (single node), Authorino Operator `quay.io/kuadrant/authorino-operator:latest`, Envoy (TLS overlay) as ext_authz, Keycloak (realm `kuadrant`); all from `kuadrant/authorino-examples`. (The bug is code-level and platform-independent; it also reproduces as a Go unit test against the existing OAuth2 test harness with a mocked 401 `{"error":...}` response.)
- Authorino version: built from `main` @ commit `7268f3a0` (`version: latest`, `dirty: false`); the same code is present in the latest release **v0.26.1**. Built with Go 1.26.4.
- Link to the Authorino `AuthConfig` resource: inline above (any `oauth2Introspection` AuthConfig triggers it once the endpoint returns a non-200 / JSON-error response).

**Additional context**

- Root cause is two missing guards in `(*OAuth2).Call`: (1) no `resp.StatusCode` check after `Do(req)`, and (2) a non-comma-ok assertion `claims["active"].(bool)`.
- Minimal fix: reject non-2xx introspection responses, and use `active, ok := claims["active"].(bool)`, returning an error when `!ok`.
- This is the same class of crash the project has fixed before; see #494 (panic on k8s SAR without `user`), #346 (crash when external OPA response misses `Content-Type`), #182 (SIGSEGV on DNS failure).
- **### I have a fix + regression test ready and am happy to open a PR.**

--- a/pkg/evaluators/identity/oauth2.go
+++ b/pkg/evaluators/identity/oauth2.go
@@ -6,6 +6,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"io"
+	"net/http"
 	"net/url"
 
 	"github.com/kuadrant/authorino/pkg/auth"
@@ -84,12 +85,28 @@ func (oauth *OAuth2) Call(pipeline auth.AuthPipeline, ctx gocontext.Context) (in
 		_ = Body.Close()
 	}(resp.Body)
 
+	// a non-200 response (e.g. invalid client credentials) is not a valid introspection
+	// result and may not carry the RFC 7662 "active" field; treat it as an error instead
+	// of attempting to parse it as a successful introspection response.
+	// The returned error only carries the status, since it is surfaced to the client (via
+	// the X-Ext-Auth-Reason header) and logged; the response body may contain data from the
+	// auth server and is only logged at debug level.
+	if resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
+		log.FromContext(ctx).WithName("oauth2").V(1).Info("token introspection request failed", "status", resp.Status, "response", string(body))
+		return nil, fmt.Errorf("token introspection request failed: %s", resp.Status)
+	}
+
 	// parse the response
 	var claims map[string]interface{}
 	if err := json.NewDecoder(resp.Body).Decode(&claims); err != nil {
 		return nil, err
 	} else {
-		if claims["active"].(bool) {
+		active, ok := claims["active"].(bool)
+		if !ok {
+			return nil, fmt.Errorf("invalid token introspection response: missing or non-boolean \"active\" field")
+		}
+		if active {
 			return claims, nil
 		} else {
 			return nil, fmt.Errorf("token is not active")

--- a/pkg/evaluators/identity/oauth2_test.go
+++ b/pkg/evaluators/identity/oauth2_test.go
@@ -22,6 +22,14 @@ func TestOAuth2Call(t *testing.T) {
 		"/introspect-inactive": func() httptest.HttpServerMockResponse {
 			return httptest.HttpServerMockResponse{Status: 200, Body: `{ "active": false }`}
 		},
+		// RFC 6749 error response returned e.g. on invalid client credentials: valid JSON, non-200, no "active".
+		"/introspect-error": func() httptest.HttpServerMockResponse {
+			return httptest.HttpServerMockResponse{Status: 401, Body: `{"error":"invalid_request","error_description":"Authentication failed."}`}
+		},
+		// 200 response that omits the "active" field.
+		"/introspect-missing-active": func() httptest.HttpServerMockResponse {
+			return httptest.HttpServerMockResponse{Status: 200, Body: `{ "sub": "1234" }`}
+		},
 	})
 	defer authServer.Close()
 
@@ -48,6 +56,23 @@ func TestOAuth2Call(t *testing.T) {
 		oauthEvaluator := NewOAuth2Identity(fmt.Sprintf("http://%v/introspect-inactive", oauthServerHost), "access_token", "client-id", "client-secret", authCredMock)
 		_, err := oauthEvaluator.Call(pipelineMock, ctx)
 		assert.Error(t, err, "token is not active")
+	}
+
+	// A non-200 introspection response (e.g. invalid client credentials) must return a clean
+	// error instead of panicking on the missing "active" field. See issue #651.
+	{
+		oauthEvaluator := NewOAuth2Identity(fmt.Sprintf("http://%v/introspect-error", oauthServerHost), "access_token", "client-id", "wrong-secret", authCredMock)
+		obj, err := oauthEvaluator.Call(pipelineMock, ctx)
+		assert.Assert(t, obj == nil)
+		assert.ErrorContains(t, err, "token introspection request failed")
+	}
+
+	// A 200 response that omits "active" must also return a clean error rather than panicking.
+	{
+		oauthEvaluator := NewOAuth2Identity(fmt.Sprintf("http://%v/introspect-missing-active", oauthServerHost), "access_token", "client-id", "client-secret", authCredMock)
+		obj, err := oauthEvaluator.Call(pipelineMock, ctx)
+		assert.Assert(t, obj == nil)
+		assert.ErrorContains(t, err, "missing or non-boolean")
 	}
 }
 

